### PR TITLE
rule: prefer-static-styles

### LINKED ISSUE #86: Rule idea: prefer adopted stylesheets
Possibly a rule which detects style tags inside templates and recommends they be moved to an adopted stylesheet (whether it be via lit or directly, we wouldn't necessarily care). 

It could have one option: a string of "prefer" or "avoid"

Someone might have a reason for preferring style tags that I'm not aware of. So we should probably give the option

If they choose "avoid" I guess we would want to detect static lit styles and adopted stylesheets to tell them not to? Maybe isn't necessary for first cut unless someone has reason you might prefer that

### PR DESCRIPTION
**Pull request by 43081j**
Fixes #86 

If `always` is set (default), it allows:
* Any assignment to an `adoptedStylesheets` property of anything (lazily since its unlikely anyone will ever name a property that which doesn't take adopted stylesheets...)
* Static styles property of a lit element class

It disallows:
* Style tags inside templates

### CODE REVIEW

File: src/rules/prefer-static-styles.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+          node.left.property.type === 'Identifier' &&
+          node.left.property.name === 'adoptedStylesheets'
+        ) {
+          context.report({
+            node,
+            messageId: 'never'
+          });
+        }
+      },
+      TaggedTemplateExpression: (node: ESTree.Node): void => {
```
Comment (stramel) [inline src/rules/prefer-static-styles.ts:86]: Shouldn't this be

```suggestion
      TaggedTemplateExpression: (node: ESTree. TaggedTemplateExpression): void => {
```

File: src/rules/prefer-static-styles.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+          node.tag.name === 'html'
+        ) {
+          const analyzer = TemplateAnalyzer.create(node);
+
+          if (prefer) {
+            analyzer.traverse({
+              enterElement(node) {
+                if (node.tagName === 'style' && node.sourceCodeLocation) {
+                  const loc = analyzer.resolveLocation(
+                    node.sourceCodeLocation.startTag,
```
Comment (stramel) [inline src/rules/prefer-static-styles.ts:99]: I think this would make more sense to highlight the entire style tag.

**Review by stramel — CHANGES_REQUESTED** [review]
This doesn't match static style getters when using option "never":
 
image

**Comment by 43081j** [comment]
updated, it now detects static styles getters

**Review by stramel — APPROVED** [review]
LGTM

**Pull request merged**